### PR TITLE
Added mod_domain_isolation

### PR DESCRIPTION
**Pull request by arcusfelis**
This PR addresses isolation.

Proposed changes include:
* isolation

### CODE REVIEW

File: src/mod_domain_isolation.erl [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    AllHosts = [Host|SubHosts],
+    [ejabberd_hooks:delete(filter_local_packet, H, ?MODULE, filter_local_packet, 10)
+     || H <- AllHosts],
+    ok.
+
+-spec filter_local_packet(Value :: fpacket() | drop) -> fpacket() | drop.
+filter_local_packet(drop) ->
+    drop;
+filter_local_packet({From = #jid{lserver = FromServer},
+                     To = #jid{lserver = ToServer}, Acc, Packet}) ->
```
Comment (NelsonVides) [inline src/mod_domain_isolation.erl:49]: Won't it happen that if we can be sure both `FromServer` and `ToServer` are the same, including subdomains, then this is allowed? In that case we could do
```suggestion
filter_local_packet({From = #jid{lserver = SameServer},
                     To = #jid{lserver = SameServer}, Acc, Packet}) ->
    {From, To, Acc, Packet};
filter_local_packet({From = #jid{lserver = FromServer},
                     To = #jid{lserver = ToServer}, Acc, Packet}) ->
```
as a shortcut path to skip extracting domains?

File: src/mod_domain_isolation.erl [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    FromHost = domain_to_host(FromServer),
+    ToHost = domain_to_host(ToServer),
+    Pass = FromHost =:= ToHost,
+    case Pass of
```
Comment (NelsonVides) [inline src/mod_domain_isolation.erl:53]:
```suggestion
    case FromHost =:= ToHost of
```
At the very least, please

**Review by NelsonVides — APPROVED** [review]
LGTM

**Comment by arcusfelis** [comment]
replaced with #3070

**Pull request merged**